### DISCUSSION TITLE
Removing expired data by IHistoryDeleteDAO in JDBC & ES has different results

### DISCUSSION THREAD

**Discussion by CzyerChen** [apache/skywalking#11906] [Q&A] [answered]
- method:  `IHistoryDeleteDAO.deleteHistory(model,TIME_BUCKET,ttl) `
- test data:  
  - fixdate now = 2024-02-21, deadline= 2024-02-18
  - tables/indices in storage: xxx-20240221, xxx-20240220, xxx-20240219, xxx-20240218, xxx-20240217, xxx-20240216
- test result:
  - JDBC:  delete xxx-20240217, xxx-20240216
      <img width="410" alt="image" src="https://github.com/apache/skywalking/assets/22817918/46d91c6a-cc40-4064-8b72-7091e6b697e6">
  - ES: delete xxx-20240218, xxx-20240217, xxx-20240216
    <img width="485" alt="image" src="https://github.com/apache/skywalking/assets/22817918/e99e4f3c-ebe7-47ec-86ed-be4615903694">
which is right? or something wrong with me?

**Comment by wu-sheng** [comment, marked as answer]
From the design perspective, `fixdate now = 2024-02-21` and `ttl=3` means data from `02-19`, `02-20`, `02-21`, and later should be kept. I think somehow, JDBC implementation keeps `02-18`.
My guessing is, that it predicts the data by current timestamp minus TTL, then gets the T - 1, meanwhile, ES implementation use the date of the current time, then minus TTL directly.

Do you want to align them?

FYI @kezhenxu94 you may know better then me.

**Reply by CzyerChen** [reply]
yes, It would be appropriate to come up with the same result.

**Reply by wu-sheng** [reply]
You could check the codes, I am not sure whether my guessing about logic is correct. But be careful, because the JDBC implementation relies on the logic to set up the max boundaries for the query range.
The fix should make sure no-time-range query works as before.

**Reply by CzyerChen** [reply]
got

**Reply by CzyerChen** [reply]
On JDBC side, count of tables for read all one more than TTL(metricsDataTTL=3/recordDataTTL=7). 
If IHistoryDeleteDAO  on JDBC side, delete tables with TLL=3(`fixdate now = 2024-02-21 and ttl=3 means data from 02-19, 02-20, 02-21`), then the count of tables for query will decrease from 4/8 to 3/7, is that ok? @kezhenxu94 May be some reasons about `ttl=3` with 4 tables for query?

**Reply by kezhenxu94** [reply]
> On JDBC side, count of tables for read all one more than TTL(metricsDataTTL=3/recordDataTTL=7). If IHistoryDeleteDAO on JDBC side, delete tables with TLL=3(`fixdate now = 2024-02-21 and ttl=3 means data from 02-19, 02-20, 02-21`), then the count of tables for query will decrease from 4/8 to 3/7, is that ok? @kezhenxu94 May be some reasons about `ttl=3` with 4 tables for query?

It's OK to me, please go ahead to make it the same as ES implementation

**Comment by CzyerChen** [comment]
Additional doubt about the `metricsDataTTL` and `recordDataTTL`, the default value in config file is 

```
recordDataTTL: ${SW_CORE_RECORD_DATA_TTL:3} # Unit is day
metricsDataTTL: ${SW_CORE_METRICS_DATA_TTL:7} # Unit is day
```

but reversed in code

<img width="666" alt="image" src="https://github.com/apache/skywalking/assets/22817918/e93410ed-aa35-4cdf-bd77-0c0e9d56d5fb">

**Reply by wu-sheng** [reply]
The codes are wrong.

**Reply by CzyerChen** [reply]
got
